lambda/matching.ml

```diff
@@ -2041,12 +2041,6 @@ let get_expr_args_record ~scopes head (arg, _mut) rem =
   make_args 0
 
 let divide_record all_labels ~scopes head ctx pm =
-  (* There is some redundancy in the expansions here, [head] is
-     expanded here and again in the matcher. It would be
-     nicer to have a type-level distinction between expanded heads
-     and non-expanded heads, to be able to reason confidently on
-     when expansions must happen. *)
-  let head = expand_record_head head in
```

Comment:
I am puzzled by the commit message.

It is true that the call to `expand_record_head` is redundant, but this does not seem to follow from recent refactorings as far as I can see: the issue referred to in the deleted comment (that we should not call `Pattern_head.arity` on a non-expanded record head pattern) is still relevant.

What happens, if that we have an invariant that ensure that record head pattern are already expanded in this case. Trying to make this invariant obvious lead me to a [not immediate refactoring](https://github.com/Octachron/ocaml/commit/13fcb0a899fd3960da9ff77e0483eace5201f195).

Reply:
I agree that the commit message is confused: `all_labels` is not available due to a recent refactoring, it was a parameter from the start. What became easy recently, I think, is to write the code so that the arity computation uses it (instead of using the provided head) in a natural way. I felt the change would have been artificial during the previous PR (but maybe I didn't look enough and I was wrong?), while it looks entirely natural in this commit. Does that sound right to you, or am I still missing something?

Reply:
Thanks for the refactoring, it looks fairly much like what I had in mind (using a different polymorphic variant for expanded record), but I hadn't gotten to implementing it yet. Would you mind rebasing at the end of the mega-PR? I would be happy to integrate it there?

Reply:
There was another arity computation in `specialize_default` that I had missed. I removed this commit.